File: src/agent/providers/bytedance-genai.ts
import type { LLMProvider } from "./base";
import type { ChatMessage, StreamChunk, ToolDefinition } from "../../types";
import { createLogger } from "../../logger";

const log = createLogger("provider:bytedance-genai");

interface ResponsesAPIInput {
  role: string;
  content: string | { type: string; text: string }[];
}

interface ResponsesAPIToolDef {
  type: "function";
  name: string;
  description: string;
  parameters: Record<string, unknown>;
}

interface ResponsesAPIRequest {
  model: string;
  input: (ResponsesAPIInput | { type: string; call_id: string; output: string })[];
  tools?: ResponsesAPIToolDef[];
  instructions?: string;
  stream?: boolean;
}

interface ResponsesAPIOutputItem {
  type: string;
  id?: string;
  content?: { type: string; text: string }[];
  role?: string;
  status?: string;
  call_id?: string;
  name?: string;
  arguments?: string;
}

interface ResponsesAPIResponse {
  id: string;
  status: string;
  output: ResponsesAPIOutputItem[];
  usage?: {
    input_tokens: number;
    output_tokens: number;
    total_tokens: number;
  };
  error?: { message: string; code: string };
}

/**
 * POST with manual redirect following that preserves method and body,
 * since standard fetch converts POST→GET on 301.
 */
async function postWithRedirect(
  url: string,
  body: string,
  headers: Record<string, string>,
  maxRedirects = 3,
): Promise<Response> {
  let currentUrl = url;
  for (let i = 0; i <= maxRedirects; i++) {
    const resp = await fetch(currentUrl, {
      method: "POST",
      headers,
      body,
      redirect: "manual",
    });
    if (resp.status === 301 || resp.status === 302 || resp.status === 307 || resp.status === 308) {
      const location = resp.headers.get("location");
      if (!location) throw new Error(`Redirect without Location header from ${currentUrl}`);
      currentUrl = new URL(location, currentUrl).toString();
      log.debug("Following redirect", { from: currentUrl, to: location });
      continue;
    }
    return resp;
  }
  throw new Error(`Too many redirects (max ${maxRedirects})`);
}

export class ByteDanceGenAIProvider implements LLMProvider {
  readonly name = "bytedance-genai";
  private baseUrl: string;
  private model: string;
  private ak: string;

  constructor(model: string, baseUrl: string, ak: string) {
    this.model = model;
    this.baseUrl = baseUrl.replace(/\/v1\/?$/, "").replace(/\/$/, "");
    this.ak = ak;
    log.info("ByteDance GenAI provider initialized", {
      model,
      baseUrl: this.baseUrl,
    });
  }

  async *chat(
    messages: ChatMessage[],
    tools?: ToolDefinition[],
    systemPrompt?: string,
  ): AsyncGenerator<StreamChunk> {
    const input: ResponsesAPIRequest["input"] = [];

    for (const msg of messages) {
      if (msg.role === "system") continue;
      if (msg.role === "tool") {
        input.push({
          type: "function_call_output",
          call_id: msg.toolCallId ?? "",
          output: msg.content,
        });
      } else if (msg.role === "assistant" && msg.toolCalls?.length) {
        if (msg.content) {
          input.push({
            role: "assistant",
            content: [{ type: "output_text", text: msg.content }],
          });
        }
        for (const tc of msg.toolCalls) {
          input.push({
            type: "function_call",
            call_id: tc.id,
            name: tc.function.name,
            arguments: tc.function.arguments,
          } as unknown as ResponsesAPIRequest["input"][number]);
        }
      } else if (msg.role === "assistant") {
        input.push({
          role: "assistant",
          content: [{ type: "output_text", text: msg.content }],
        });
      } else {
        input.push({
          role: msg.role,
          content: [{ type: "input_text", text: msg.content }],
        });
      }
    }

    const reqTools: ResponsesAPIToolDef[] | undefined = tools?.length
      ? tools.map((t) => ({
          type: "function" as const,
          name: t.name,
          description: t.description,
          parameters: t.parameters,
        }))
      : undefined;

    const reqBody: ResponsesAPIRequest = {
      model: this.model,
      input,
      stream: false,
    };
    if (systemPrompt) {
      reqBody.instructions = systemPrompt;
    }
    if (reqTools?.length) {
      reqBody.tools = reqTools;
    }

    const url = `${this.baseUrl}/responses?ak=${encodeURIComponent(this.ak)}`;
    const bodyStr = JSON.stringify(reqBody);

    try {
      const resp = await postWithRedirect(url, bodyStr, {
        "Content-Type": "application/json",
      });

      if (!resp.ok) {
        const errText = await resp.text();
        log.error("API request failed", { status: resp.status, body: errText });
        yield { type: "error", error: `API error ${resp.status}: ${errText}` };
        return;
      }

      const data = (await resp.json()) as ResponsesAPIResponse;

      if (data.error) {
        yield { type: "error", error: data.error.message };
        return;
      }

      for (const item of data.output) {
        if (item.type === "message" && item.content) {
          for (const part of item.content) {
            if (part.type === "output_text" && part.text) {
              yield { type: "text", content: part.text };
            }
          }
        } else if (item.type === "function_call" && item.name) {
          yield {
            type: "tool_call",
            toolCall: {
              id: item.call_id ?? item.id ?? `tc_${Date.now()}`,
              type: "function",
              function: {
                name: item.name,
                arguments: item.arguments ?? "{}",
              },
            },
          };
        }
      }

      yield {
        type: "done",
        usage: data.usage
          ? {
              promptTokens: data.usage.input_tokens,
              completionTokens: data.usage.output_tokens,
              totalTokens: data.usage.total_tokens,
            }
          : undefined,
      };
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      log.error("ByteDance GenAI request failed", { error: message });
      yield { type: "error", error: message };
    }
  }
}

File: src/models/bytedance_genai.py
"""ByteDance GenAI Responses API provider, matching TS providers/bytedance-genai.ts."""

from __future__ import annotations

import json
import re
import time
from typing import AsyncGenerator

import httpx

from src.types import (
    ChatMessage,
    StreamChunk,
    TokenUsage,
    ToolCall,
    ToolCallFunction,
    ToolDefinition,
)
from src.utils.logger import create_logger

log = create_logger("provider:bytedance-genai")


async def _post_with_redirect(
    client: httpx.AsyncClient,
    url: str,
    body: str,
    headers: dict[str, str],
    max_redirects: int = 3,
) -> httpx.Response:
    current_url = url
    for _ in range(max_redirects + 1):
        resp = await client.post(
            current_url,
            content=body,
            headers=headers,
            follow_redirects=False,
        )
        if resp.status_code in (301, 302, 307, 308):
            location = resp.headers.get("location")
            if not location:
                raise RuntimeError(f"Redirect without Location header from {current_url}")
            current_url = location
            log.debug("Following redirect", {"to": location})
            continue
        return resp
    raise RuntimeError(f"Too many redirects (max {max_redirects})")


class ByteDanceGenAIProvider:
    def __init__(self, model: str, base_url: str, ak: str) -> None:
        self._model = model
        self._base_url = re.sub(r"/v1/?$", "", base_url.rstrip("/"))
        self._ak = ak
        log.info("ByteDance GenAI provider initialized", {"model": model, "baseUrl": self._base_url})

    @property
    def name(self) -> str:
        return "bytedance-genai"

    async def chat(
        self,
        messages: list[ChatMessage],
        tools: list[ToolDefinition] | None = None,
        system_prompt: str | None = None,
    ) -> AsyncGenerator[StreamChunk, None]:
        input_items: list[dict] = []

        for msg in messages:
            if msg.role == "system":
                continue
            if msg.role == "tool":
                input_items.append({
                    "type": "function_call_output",
                    "call_id": msg.tool_call_id or "",
                    "output": msg.content,
                })
            elif msg.role == "assistant" and msg.tool_calls:
                if msg.content:
                    input_items.append({
                        "role": "assistant",
                        "content": [{"type": "output_text", "text": msg.content}],
                    })
                for tc in msg.tool_calls:
                    input_items.append({
                        "type": "function_call",
                        "call_id": tc.id,
                        "name": tc.function.name,
                        "arguments": tc.function.arguments,
                    })
            elif msg.role == "assistant":
                input_items.append({
                    "role": "assistant",
                    "content": [{"type": "output_text", "text": msg.content}],
                })
            else:
                input_items.append({
                    "role": msg.role,
                    "content": [{"type": "input_text", "text": msg.content}],
                })

        req_body: dict = {
            "model": self._model,
            "input": input_items,
            "stream": False,
        }
        if system_prompt:
            req_body["instructions"] = system_prompt
        if tools:
            req_body["tools"] = [
                {
                    "type": "function",
                    "name": t.name,
                    "description": t.description,
                    "parameters": t.parameters,
                }
                for t in tools
            ]

        url = f"{self._base_url}/responses?ak={self._ak}"
        body_str = json.dumps(req_body)

        try:
            async with httpx.AsyncClient(timeout=60) as client:
                resp = await _post_with_redirect(
                    client, url, body_str, {"Content-Type": "application/json"}
                )

            if resp.status_code != 200:
                err_text = resp.text
                log.error("API request failed", {"status": resp.status_code, "body": err_text})
                yield StreamChunk(type="error", error=f"API error {resp.status_code}: {err_text}")
                return

            data = resp.json()
            if data.get("error"):
                yield StreamChunk(type="error", error=data["error"]["message"])
                return

            for item in data.get("output", []):
                if item.get("type") == "message" and item.get("content"):
                    for part in item["content"]:
                        if part.get("type") == "output_text" and part.get("text"):
                            yield StreamChunk(type="text", content=part["text"])
                elif item.get("type") == "function_call" and item.get("name"):
                    yield StreamChunk(
                        type="tool_call",
                        tool_call=ToolCall(
                            id=item.get("call_id") or item.get("id") or f"tc_{int(time.time()*1000)}",
                            function=ToolCallFunction(
                                name=item["name"],
                                arguments=item.get("arguments", "{}"),
                            ),
                        ),
                    )

            usage = data.get("usage")
            yield StreamChunk(
                type="done",
                usage=TokenUsage(
                    prompt_tokens=usage["input_tokens"],
                    completion_tokens=usage["output_tokens"],
                    total_tokens=usage["total_tokens"],
                ) if usage else None,
            )

        except Exception as e:
            log.error("ByteDance GenAI request failed", {"error": str(e)})
            yield StreamChunk(type="error", error=str(e))
